Add iss, aud validations, and extend validationError

axos88 opened the pull request:
This PR drops the String result from the validation errors, and replaces them with more fine-grained types / enum varians.

It also refactors the validation of the token and adds iss and aud validations, and makes it possible to require the presence of other RFC-mandated fields.

lawliet89 commented:
Thanks for the PR. Let me have some time to look over it. In the meantime, you might want to fix a compile error: https://travis-ci.org/lawliet89/biscuit/jobs/406826463#L663-L669

axos88 commented:
Whoops, forgot to push the latest commit.

lawliet89 commented:
I am going to push a commit to your branch after running `cargo fmt` so that there is less "formatting" comments.

lawliet89 commented on src/lib.rs:
```diff
@@ -851,103 +854,272 @@ pub struct RegisteredClaims {
+    /// Whether the `aud` or `Audience` field is required
+    /// Token must include an audience with the value of the parameter
+    pub audience: Validation<StringOrUri>,
+    //    pub subject_validated: Validation<Box<FnOnce(String) -> bool>>,
```
I think we should leave all non RFC validation "outside" the library so that users can add their bespoke validation on top.

lawliet89 commented on src/lib.rs:
```diff
@@ -851,103 +854,272 @@ pub struct RegisteredClaims {
+            temporal_options: Default::default(),
+            audience: Default::default(),
+            issuer: Default::default(),
+            //            subject_validated: Default::default(),
```
Should remove if unused.

lawliet89 commented on src/lib.rs:
```diff
@@ -851,103 +854,272 @@ pub struct RegisteredClaims {
-    /// are required, and they will pass validation if they are missing.
-    pub fn validate_times(&self, options: Option<TemporalValidationOptions>) -> Result<(), ValidationError> {
-        let options = options.unwrap_or_default();
+    ///Validates that the token contains the claims defined as required
```
Space after `///`

lawliet89 commented on src/lib.rs:
```diff
@@ -851,103 +854,272 @@ pub struct RegisteredClaims {
+    }
 
-        Ok(())
+    ///Validates that if the token has an `exp` claim, it has not passed.
```
Space after `///`

lawliet89 commented on src/lib.rs:
```diff
@@ -851,103 +854,272 @@ pub struct RegisteredClaims {
+    pub claim_presence_options: ClaimPresenceOptions,
+
+    ///Define how to validate temporal claims
+    pub temporal_options: TemporalOptions,
```
Instead of having `TemporalOptions` as a field, can we have it so that it is part of the `T` for the `Validation` generics for `issued_at`, `not_before`, and `expiry`? More granularity.

axos88 replied:
That was the way I implemented it at first, but then I realized there isn't really a use case for using different now() or epsilon for different claims

lawliet89 commented on src/lib.rs:
```diff
@@ -851,103 +854,272 @@ pub struct RegisteredClaims {
     /// Whether the `iat` or `Issued At` field is required
-    pub issued_at_required: bool,
+    /// Parameter shows the maximum age of a token to be accepted.
+    pub issued_at: Validation<Duration>,
```
As above, let's have this be `Validation<(TemporalOptions, Duration)>` instead.

lawliet89 commented on src/lib.rs:
```diff
@@ -851,103 +854,272 @@ pub struct RegisteredClaims {
+    pub issued_at: Validation<Duration>,
     /// Whether the `nbf` or `Not Before` field is required
-    pub not_before_required: bool,
+    pub not_before: Validation<()>,
```
Let this be `Validation<TemporalOptions>` instead.

lawliet89 commented on src/lib.rs:
```diff
@@ -851,103 +854,272 @@ pub struct RegisteredClaims {
-    pub epsilon: Option<std::time::Duration>,
-    /// Specify a time to use in temporal validation instead of `Now`.
-    pub now: Option<DateTime<Utc>>,
+    pub expiry: Validation<()>,
```
Let this be `Validation<TemporalOptions>` instead.

lawliet89 commented on src/lib.rs:
```diff
@@ -851,103 +854,272 @@ pub struct RegisteredClaims {
-        // `time::OutOfRangeError`. We don't want to put `time` as a dependent crate just to `impl From` for this...
-        // So I am just going to `map_err`.
-        use std::error::Error;
+    ///Validates that if the token has an `nbf` claim, it has passed.
```
Space after `///`

lawliet89 commented on src/lib.rs:
```diff
@@ -851,103 +854,272 @@ pub struct RegisteredClaims {
-        // `time::OutOfRangeError`. We don't want to put `time` as a dependent crate just to `impl From` for this...
-        // So I am just going to `map_err`.
-        use std::error::Error;
+    ///Validates that if the token has an `iat` claim, it is not in the future and not older than the Duration
```
Space after `///`

lawliet89 commented on src/lib.rs:
```diff
@@ -851,103 +854,272 @@ pub struct RegisteredClaims {
 
-        let e = chrono::Duration::from_std(e).map_err(|e| ValidationError::TemporalError(e.description().to_string()))?;
-        Ok(a.signed_duration_since(b) <= e)
+    ///Validates that if the token has an `aud` claim, it contains an entry which matches the expected audience
```
Space after `///`

lawliet89 commented on src/lib.rs:
```diff
@@ -851,103 +854,272 @@ pub struct RegisteredClaims {
+        }
+    }
+
+    ///Validates that if the token has an `iss` claim, it matches the expected issuer
```
Space after `///`

lawliet89 commented on src/lib.rs:
```diff
@@ -851,103 +854,272 @@ pub struct RegisteredClaims {
+        match validation {
+            Validation::Ignored => Ok(()),
+            Validation::Validate(temporal_options) => {
+                let now = temporal_options.now.unwrap_or(Utc::now());
```
Use [`unwrap_or_else`](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or_else) for lazy evaluation.

axos88 replied:
Is this such an expensive operation that it's worth it?

lawliet89 commented on src/lib.rs:
```diff
@@ -851,103 +854,272 @@ pub struct RegisteredClaims {
+        match validation {
+            Validation::Ignored => Ok(()),
+            Validation::Validate(temporal_options) => {
+                let now = temporal_options.now.unwrap_or(Utc::now());
```
Use [`unwrap_or_else`](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or_else) for lazy evaluation.

lawliet89 commented on src/lib.rs:
```diff
@@ -851,103 +854,272 @@ pub struct RegisteredClaims {
+        match validation {
+            Validation::Ignored => Ok(()),
+            Validation::Validate((max_age, temporal_options)) => {
+                let now = temporal_options.now.unwrap_or(Utc::now());
```
Use [`unwrap_or_else`](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or_else) for lazy evaluation.

lawliet89 commented on src/lib.rs:
```diff
@@ -851,103 +854,272 @@ pub struct RegisteredClaims {
+    ///Define how to validate temporal claims
+    pub temporal_options: TemporalOptions,
+
     /// Whether the `iat` or `Issued At` field is required
```
This should now read something like 

    Validation options for `iat` or `Issued At` field if present

Let's also add "tips" for the user in case they don't want to use the maximum age (i.e. set it to `Duration::max_value()`)

lawliet89 commented on src/lib.rs:
```diff
@@ -851,103 +854,272 @@ pub struct RegisteredClaims {
-    pub issued_at_required: bool,
+    /// Parameter shows the maximum age of a token to be accepted.
+    pub issued_at: Validation<Duration>,
     /// Whether the `nbf` or `Not Before` field is required
```
This should now read something like 

    Validation options for `nbf` or `Not Before` field if present

lawliet89 commented on src/lib.rs:
```diff
@@ -851,103 +854,272 @@ pub struct RegisteredClaims {
     /// Whether the `nbf` or `Not Before` field is required
-    pub not_before_required: bool,
+    pub not_before: Validation<()>,
     /// Whether the `exp` or `Expiry` field is required
```
This should now read something like 

    Validation options for `exp` or `Expiry` field if present

lawliet89 commented on src/lib.rs:
```diff
@@ -851,103 +854,272 @@ pub struct RegisteredClaims {
-    pub now: Option<DateTime<Utc>>,
+    pub expiry: Validation<()>,
+
+    /// Whether the `iss` or `Issuer` field is required
```
This should now read something like 

    Validation options for `iss` or `Issuer` field if present

lawliet89 commented on src/lib.rs:
```diff
@@ -851,103 +854,272 @@ pub struct RegisteredClaims {
+    /// Parameter must match the issuer in the token exactly.
+    pub issuer: Validation<StringOrUri>,
+
+    /// Whether the `aud` or `Audience` field is required
```
This should now read something like 

    Validation options for `aud` or `Audience` field if present

lawliet89 commented on src/lib.rs:
```diff
@@ -851,103 +854,272 @@ pub struct RegisteredClaims {
+/// `claim_presence_options`.
+///
+/// By default only expiry-related claims are validated (namely `exp`, `nbf`, `iat`).
+///
```
Extra line breakes -- please remove.

lawliet89 commented on src/lib.rs:
```diff
@@ -851,103 +854,272 @@ pub struct RegisteredClaims {
+
+impl Default for ValidationOptions {
+    fn default() -> Self {
+        ValidationOptions {
```
This whole block can be simplified:

```rust
ValidationOptions {
  issued_at: Validation::Validate((Default::default(), Duration::max_value())),
  ..Default::default()
}
```

axos88 replied:
Well `expiry` and `not_before` also have non-default values, and I wanted the full list here so that anyone browsing through the code will see instantly what the default values are without having to go deeper

lawliet89 commented on src/lib.rs:
```diff
@@ -851,103 +854,272 @@ pub struct RegisteredClaims {
+/// If a claim is missing, it passes validation unless the claim is marked as required within the
+/// `claim_presence_options`.
+///
+/// By default only expiry-related claims are validated (namely `exp`, `nbf`, `iat`).
```
Should document more on the default.

   By default, no claims are required. If there are present, only expiry-related claims are validated 
   (namely `exp`, `nbf`, `iat`) with zero epsilon and maximum age duration.

axos88 replied:
Done

lawliet89 replied:
Thanks for adding them. I think the lines are now repeated?

Line 900 and 903.

lawliet89 commented on src/jwe.rs:
```diff
@@ -601,8 +601,8 @@ where
     /// are required, and they will pass validation if they are missing.
-    pub fn validate_times(&self, options: Option<::TemporalValidationOptions>) -> Result<(), Error> {
-        Ok(self.payload()?.registered.validate_times(options)?)
+    pub fn validate_times(&self, options: ::ValidationOptions) -> Result<(), Error> {
```
This method should be renamed to `validate`.

axos88 commented on src/helpers/validation.rs:
```diff
@@ -0,0 +1,37 @@
+#[derive(Debug, Eq, PartialEq, Clone, Copy)]
+/// Defines wether a claim is validated or not
```
What funky `[...]` syntax?

axos88 replied:
Ah gotcha.

axos88 commented on src/helpers/validation.rs:
```diff
@@ -0,0 +1,38 @@
+    Validate(T),
+}
+
+// This doesn't compile currently
```
It does compile on nightly though with `#![feature(specialization)]`. I will remove it since it doesn't add much

axos88 commented:
Thank you for your detailed review @lawliet89, sorry for the ton of whitespace and documentation errors and that it took so much time to fix them, i had my hands full this week.

Please review again, hopefully I addressed your concerns :)

axos88 commented:
Huh. Must've been some weird git merge issue, had a few conflicts - i had to revert the ECDSA commit in my fork because actix-web stopped working with that (due to the upgrade of the ring lib)

lawliet89 commented:
Good point. I'll need to get around to updating `ring` dependency. Probably middle of next month

lawliet89 commented:
Thanks for your contribution.
